### ISSUE TITLE
Optimize ad serving based upon local user browser activity

### ISSUE THREAD

**Issue by tmancey** [brave/brave-browser#14045]
Optimize ad serving based upon a users’ activity by locally monitoring the following user-initiated events:
```
- Started ads
- Browser window did become active
- Browser window did enter the background
- Clicked back or forward navigation buttons
- Clicked a bookmark
- Clicked the home page button
- Clicked a link
- Clicked the reload button
- Closed a tab
- Focused on an existing tab
- Opened a link from an external application
- Opened a new tab
- Played media
- Stopped playing media
- Submitted a form
- Updated a tab
- Typed in the URL bar
- Typed in the URL bar and selected an entry that did not look like a URL
- Used the address bar
- New navigation
- URL was generated from a replaceable keyword other than the default search provider
- Tab to search
```
user-initiated events should be transient (in memory and never persisted), we should only store an `id` and `timestamp` and never leave the users' device.

We calculate the users’ activity score from events triggered by the user within the specified `time_window`. If the score is >= the `threshold`, and the user is within the various parameters below, we can assume we have gained the users’ attention.

`triggers`: A list of event triggers
`time_window`: The timeframe to consider if a user was active
`threshold`: The threshold to exceed to be considered active
`idle_time_threshold`: Minimum idle time before an ad maybe served. Defaults to `15`
`maximum_idle_time`: Maximum idle time to consider serving an ad. Defaults to `0` (infinite)
`should_detect_was_locked`: Whether to detect if a user is unidle straight after unlocking their device. Defaults to `false`

The above parameters should be configurable via the server to allow Brave to tweak the users' experience.

**Comment by GeetaSarvadnya** [comment]
Verification is in-progress

```

Brave | 1.23.68 Chromium: 90.0.4430.51 (Official Build) (64-bit)
-- | --
Revision | 32e5fa33a31641bded70a90e60121060691e7125-refs/branch-heads/4430@{#927}
OS | Windows 10 OS Version 2004 (Build 19041.867)

```
- Verified the test plan from https://github.com/brave/brave-core/pull/7881

<details>
<summary>Case 1a: Insufficient activity with default triggers</summary>

</details>

**Comment by srirambv** [comment]
Verification passed on OnePlus 6T with Android 10 running 1.23.69 x64 build
<details>
<summary>Verified test plan from <a href=https://github.com/brave/internal/issues/763>brave/internal/#763</a></summary>
<br>

<li> Verified ads show up as expected as per setting. More info <a href=https://bravesoftware.slack.com/archives/C43LW5U2H/p1618328349048400?thread_ts=1618323511.047300&cid=C43LW5U2H>here</a> </li>

<br>
<details>
<summary><b>User Activity - Default values - kInitializedAds </b> </summary> 
<br> - Verified when ads is initialized event 00 is triggered 


``` 
2021-04-13 03:58:22.746 18371-18371/? V/chromium: [VERBOSE1:ads_impl.cc(521)] Successfully initialized ads 2021-04-13 03:58:22.746 18371-18371/? V/chromium: [VERBOSE6:user_activity.cc(41)] Triggered event: 00 (0:2:3600 s) 
``` 
</details>

<details>
<summary><b>User Activity - Default values - kBrowserDidBecomeActive</b></summary> 
<br> - Verified when browser becomes active event 01 is triggered with value of 2 to 3 

``` 
2021-04-13 03:58:55.235 18371-18371/? V/chromium: [VERBOSE1:browser_manager.cc(78)] Browser did become active 2021-04-13 03:58:55.235 18371-18371/? V/chromium: [VERBOSE6:user_activity.cc(41)] Triggered event: 01 (2:2:3600 s) 2021-04-13 04:33:04.549 18371-18371/? V/chromium: [VERBOSE1:browser_manager.cc(78)] Browser did become active 2021-04-13 04:33:04.549 18371-18371/? V/chromium: [VERBOSE6:user_activity.cc(41)] Triggered event: 01 (3:2:3600 s) 
``` 

</details>
<details>
<summary><b>User Activity - Default values - kBrowserDidEnterBackground</b></summary> 
<br> - Verified when browser goes to background event 02 is triggered with value of 2.5 to 3.5 

``` 
2021-04-13 04:08:55.253 18371-18371/? V/chromium: [VERBOSE1:browser_manager.cc(91)] Browser did enter background 2021-04-13 04:08:55.253 18371-18371/? V/chromium: [VERBOSE6:user_activity.cc(41)] Triggered event: 02 (2.5:2:3600 s) 2021-04-13 04:33:04.608 18371-18371/? V/chromium: [VERBOSE1:browser_manager.cc(91)] Browser did enter background 2021-04-13 04:33:04.608 18371-18371/? V/chromium: [VERBOSE6:user_activity.cc(41)] Triggered event: 02 (3.5:2:3600 s) 
``` 

</details>
<details>
<summary><b>User Activity - Default values - kClickedLink</b></summary> 
<br> - Verified when a link is clicked event 06 is triggered with a valule of 4 

``` 
2021-04-13 06:26:56.991 18371-18371/? V/chromium: [VERBOSE6:user_activity.cc(41)] Triggered event: 06 (4:2:3600 s) 
``` 

</details>
<details>
<summary><b>User Activity - Default values - kClosedTab</b></summary> 
<br> - Verified when a tab is closed event 08 is triggered 

``` 
2021-04-13 07:16:53.562 22171-22171/? V/chromium: [VERBOSE2:tab_manager.cc(113)] Tab id 2 was closed 2021-04-13 07:16:53.565 22171-22171/? V/chromium: [VERBOSE6:user_activity.cc(41)] Triggered event: 08 (4:2:3600 s) 
``` 

</details>
<details>
<summary><b>User Activity - Default values - kFocusedOnExistingTab</b></summary> 
<br> - Verified when focus is set to an existing tab event 09 is triggered and value changes to 4.5 

``` 
2021-04-13 06:28:34.480 18371-18371/? V/chromium: [VERBOSE6:user_activity.cc(41)] Triggered event: 09 (4.5:2:3600 s) 
``` 

</details>
<details>
<summary><b>User Activity - Default values - kNewNavigation</b></summary> 
<br> - Verified when a new page is navigated event 0B is triggered with score of 4 

``` 
2021-04-13 06:26:56.991 18371-18371/? V/chromium: [VERBOSE6:user_activity.cc(41)] Triggered event: 0B (4:2:3600 s)
``` 

</details>
<details>
<summary><b>User Activity - Default values - kOpenedNewTab</b></summary> 
<br> - Verified when a new tab is opened event 0D is triggered with a score of 6 

``` 
2021-04-13 04:40:53.178 18371-18371/? V/chromium: [VERBOSE2:tab_manager.cc(104)] Opened a new tab with id 3 2021-04-13 04:40:53.178 18371-18371/? V/chromium: [VERBOSE6:user_activity.cc(41)] Triggered event: 0D (6:2:3600 s) 
``` 

</details>
<details>
<summary><b>User Activity - Default values - kPlayedMedia</b></summary> 
<br> - Verified when a media is playing triggers event 0E 

``` 
2021-04-13 07:18:49.537 22171-22171/? V/chromium: [VERBOSE2:tab_manager.cc(125)] Tab id 12 is playing media 2021-04-13 07:18:49.540 22171-22171/? V/chromium: [VERBOSE6:user_activity.cc(41)] Triggered event: 0E (11:2:3600 s) 
``` 

</details>
<details>
<summary><b>User Activity - Default values - kStoppedPlayingMedia</b></summary> 
<br> - Verified when media stops playing event 0F is triggered 

``` 
2021-04-13 07:18:52.487 22171-22171/? V/chromium: [VERBOSE2:tab_manager.cc(137)] Tab id 12 stopped playing media 2021-04-13 07:18:52.490 22171-22171/? V/chromium: [VERBOSE6:user_activity.cc(41)] Triggered event: 0F (11:2:3600 s) 
``` 

</details>
<details>
<summary><b>User Activity - Default values - kTabUpdated</b></summary> 
<br> - Verified when browser opens a new tab event 11 is triggered with a score of 5 

``` 
2021-04-13 04:34:48.040 18371-18371/? V/chromium: [VERBOSE2:tab_manager.cc(77)] Tab id 2 was updated 2021-04-13 04:34:48.041 18371-18371/? V/chromium: [VERBOSE6:user_activity.cc(41)] Triggered event: 11 (5:2:3600 s) 
``` 

</details>
<details>
<summary><b>User Activity - Default values - kBrowserWindowIsActive</b></summary> 
<br> - Verified when browser becomes active event 16 is triggered with score of 0.5 

``` 
2021-04-13 06:34:43.687 22171-22171/? V/chromium: [VERBOSE1:browser_manager.cc(48)] Browser window is active 2021-04-13 06:34:43.690 22171-22171/? V/chromium: [VERBOSE6:user_activity.cc(41)] Triggered event: 16 (0.5:2:3600 s) 
``` 

</details>
</details>

---------
Verification passed on Samsung Tab A with Android 10 running 1.23.69 x64 build
<details>
<summary>Verified test plan from <a href=https://github.com/brave/internal/issues/763>brave/internal/#763</a></summary>
<br>

<li> Verified ads show up as expected as per setting. More info <a href=https://bravesoftware.slack.com/archives/C43LW5U2H/p1618328349048400?thread_ts=1618323511.047300&cid=C43LW5U2H>here</a> </li>

</details>
